Component requirements shall need either Default or to be a unit struct.

coolcatcoder opened the pull request:
# Objective

- Removes the need to derive Default on a unit struct that wishes to be a requirement of a component.
- Fixes #18859

## Solution

- Auto deref specialisation. Both const and non-const.

## Testing

- We must check that requirements with default are allowed, and ones without default are denied unless they so happen to be a unit struct.

coolcatcoder committed:
A few failed attempts.

Not certain how small you want each commit to be? I'll just commit every now and again I suppose.

coolcatcoder committed:
rock and a hard place

 Continuous integration complains that my let binding has unit value, but if I remove it, then it claims that my code is a path statement with no effect.

coolcatcoder committed:
Switch constant to const fn!

This gets rid of the continuous integration errors.

coolcatcoder committed:
Nothing needs default! Ah!

Because this is inside a trait, I need to use a slightly different approach. Whoops. Should be fixed next commit.

coolcatcoder committed:
Correctly deny invalid requirements.

Had to switch to const deref specialisation, but it caused the error message to improve, which is quite nice.

coolcatcoder committed:
extra assurance

Added a check to make 100% sure that only 0 sized types don't use default.

coolcatcoder committed:
Remove random hidden space.

coolcatcoder committed:
Switch to a compile fail test!

It doesn't work yet, but it is along the right track.
Also why are there so many clippy issues? I didn't even touch the code they are complaining about!

coolcatcoder committed:
Fixed the compile fail!

coolcatcoder committed:
clean up

benfrankel commented on .cargo/config_fast_builds.toml:
This was probably not meant to be deleted.

coolcatcoder replied:
I wasn't exactly sure how to handle the fast build toml. I don't yet know enough about git to know how to exclude my changes to that directory. I suppose I'll just revert those files back to normal at the end.

benfrankel commented on crates/bevy_ecs/src/component.rs:
```diff
@@ -37,6 +37,72 @@ use disqualified::ShortName;
+}
+
+/// Checks if a component's requirement is valid.
+/// It is valid if it has Default, or if it has a size of 0 and is therefore a unit struct.
```
Technically not all ZSTs are unit structs. For example, `struct Foo { x: () }` is a ZST but not a unit struct. Also empty types like `!` and `enum Bar {}` are ZSTs.

coolcatcoder replied:
Yeah. I am worried about that, but at the same time, it seems unlikely for such structs to be marked as a component. Are there any foot guns I haven't noticed?

coolcatcoder replied:
If a struct has a single private field that is a zst, so that people cannot construct it, then this bypasses it. Yikes. Unlikely to be a problem, but still not great.
We could solve this by making the component derive implement a UnitStruct trait when it encounters a unit struct.

benfrankel commented on crates/bevy_ecs/compile_fail/tests/ui/non_unit_without_default_required_components.rs:
```diff
@@ -0,0 +1,10 @@
+use bevy_ecs::prelude::*;
+
+#[derive(Component)]
+
+struct A{bad:i32}
```
```suggestion
#[derive(Component)]
struct A {
    bad: i32,
}
```

benfrankel commented on crates/bevy_ecs/compile_fail/tests/ui/non_unit_without_default_required_components.rs:
```diff
@@ -0,0 +1,10 @@
+//~v E0080
+#[derive(Component)]
+#[require(A)]
+struct B;
```
```suggestion
struct B;

```

benfrankel reviewed (changes requested):
I'm curious about the motivation for this feature, as opposed to deriving `Default` on the required component. Is this meant for components that are defined in external crates?

EDIT: I see there's a linked issue but this seems to be an open question there as well.

coolcatcoder committed:
Update crates/bevy_ecs/compile_fail/tests/ui/non_unit_without_default_required_components.rs

Thanks benfrankel! That was left over from when I accidentally put the error checker in the wrong place...

Co-authored-by: Ben Frankel <ben.frankel7@gmail.com>

coolcatcoder committed:
Update crates/bevy_ecs/compile_fail/tests/ui/non_unit_without_default_required_components.rs

Thanks benfrankel!

Co-authored-by: Ben Frankel <ben.frankel7@gmail.com>

coolcatcoder commented:
> I'm curious about the motivation for this feature, as opposed to deriving `Default` on the required component. Is this meant for components that are defined in external crates?
> 
> EDIT: I see there's a linked issue but this seems to be an open question there as well.

I have no idea why this would be wanted. I simply saw an issue that it was claimed was not possible to solve, and as such I just had to solve it. (It was fun! Although, I did spend 90% of my time fighting with continuous integration, because I had never used it before!)

alice-i-cecile commented:
> > Technically not all ZSTs are unit structs. For example, `struct Foo { x: () }` is a ZST but not a unit struct. Also empty types (like `!` and `enum Bar {}`) and functions are ZSTs.

This makes me very nervous. I don't think that the value to risk / complexity is worth it here. I've marked the upstream issue Nominated-to-Close, and think we should do the same here.
